### ISSUE TITLE
Find a way to set a default global per-test timeout for Java Java unit tests

### ISSUE THREAD

**Issue by damccorm** [apache/beam#21377]
Apparently java unit tests don't have timeouts. And as far as I can tell, Junit lets you set timeouts for individual tests and test classes, but provides no easy way to set a global timeout across multiple classes.

Imported from Jira [BEAM-13523](https://issues.apache.org/jira/browse/BEAM-13523). Original Jira may contain additional context.
Reported by: ibzib.

**Comment by gpt3ch** [comment]
Hello @damccorm , I started looking at this issue. After reading more about this, a way to implement a global timeouts for the unit tests is to create a base test class and specify a Timeout rule in it, then make all the test classes inherit the base test class, this will make all the test cases timeout if they cross the threshold without having to specify @Timeout annotation over each test method. Is this an acceptable solution?

**Comment by damccorm** [comment]
Hey @gpt3ch! That definitely seems reasonable to me, but I also don't have a ton of context on the Java test suites (I just migrated the issue to GitHub, this was originally reported by @ibzib).

Tagging a few folks who might have more context here in case they have thoughts: @ibzib, @klk, @kileys, or @y1chi

**Comment by sulthan309** [comment]
Hello @damccorm, 'Global declarative Timeout' feature is implemented in Junit from Junit 5.5 version where we can set the global timeout for all the TCs.
Documentation : https://junit.org/junit5/docs/5.5.0/user-guide/index.html#writing-tests-declarative-timeouts (section **2.19.2. Default Timeouts**)

**Comment by Ashutosh2512** [comment]
Hi @damccorm,

Is the issue still relevant or the solution provided by @sulthan309 is acceptable, given it is an experimental feature from JUnit ? If it is relevant, I would like to pick it up.

**Comment by arnavarora2004** [comment]
.take-issue

**Comment by arnavarora2004** [comment]
Hey @damccorm !

Just my two cents on this ticket based on what's been given so far:


> Hello [@damccorm](https://github.com/damccorm), 'Global declarative Timeout' feature is implemented in Junit from Junit 5.5 version where we can set the global timeout for all the TCs. Documentation : https://junit.org/junit5/docs/5.5.0/user-guide/index.html#writing-tests-declarative-timeouts (section **2.19.2. Default Timeouts**)


After looking at  [This](https://github.com/apache/beam/blob/1de0d1f642d1c03bfc45a13e48fe967dab133324/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L834) link you can see that the Junit current version is 4.13.1, which does not allow for  configuration parameters changes, 

Doing some more in depth research I found:

- in Maven (which to my understanding is Beam's dedicated test suite runner for Junit tests) or Gradle test tasks, configure a timeout at the suite level, which will encompass all tests in the suite, so the [assembly.xml](https://github.com/apache/beam/blob/master/assembly.xml) file in root or [pom.xml](https://github.com/apache/beam/blob/master/sdks/java/maven-archetypes/starter/pom.xml) files a bit deeper for Maven  and the build.gradle files for Gradle

 - create a base test class and specify a Timeout rule in it, then make all the test classes inherit the base test class (if the test classes already inherit a skeleton base test class simply adding an @Rule to the base test class file will do the trick)

- upgrade Junit to 5+ and then go to configuration parameters and add default Timeout (involves major changes if methods and stuff depreciated) 
- I feel there's a way to add timeouts in the [gradlew](https://github.com/apache/beam/blob/master/gradlew) file in root, trying to do more research on that


What do we think is best considering all this info?

**Comment by damccorm** [comment]
cc/ @Abacn who may have ideas.

> upgrade Junit to 5+ and then go to configuration parameters and add default Timeout (involves major changes if methods and stuff depreciated)

I think it is worth trying this and seeing how hard it is - IMO this is probably the "right" solution since it is minimally invasive, and upgrading Junit is probably a generally good thing for the project. I'd recommend trying the upgrade, opening a PR, and seeing what fails in our CI. If it is small/easy to fix, lets do it, if not we can look for other paths forward

**Comment by Abacn** [comment]
Yeah I agree Junit 5 it's the right path. junit 4 still on java 7 and never make use of more modern java feature.

**Comment by arnavarora2004** [comment]
One problem I am seeing right off the bat @Abacn @damccorm is that the Maven version is 2 and does not support Junit 5, 
Leads to this error:
"Could not find junit:junit:5.12.2.
     Searched in the following locations:
       - file:/runner/_work/beam/beam/sdks/java/core/offline-repository/junit/junit/5.12.2/junit-5.12.2.pom
       - https://repo.maven.apache.org/maven2/junit/junit/5.12.2/junit-5.12.2.pom"

which fails a few of the checks in the CI, from what I understand even though we may not use Maven as the primary test suite runner it still causes dependency issues so should I go ahead and keep working up this chain of dependency errors or stop? Not sure how far up it will go but if Maven isn't even used I could simply get rid of the tests/files using it?

just let me know,

**Comment by Abacn** [comment]
there is no `junit:junit:5.x` in https://mvnrepository.com/artifact/junit/junit. junit 5 changed artifact names

> JUnit 4 ... A single jar library contains the entire framework.
> In JUnit 5, we get more granularity and can import only what’s necessary.

(ref - https://www.baeldung.com/junit-5-migration)

You'd need to use [org.junit.jupiter](https://mvnrepository.com/artifact/org.junit.jupiter) and [junit-jupiter-api](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api)

**Comment by arnavarora2004** [comment]
[link to CI checks](https://github.com/apache/beam/pull/34958) 

Let me know what you guys think, from the looks of it I just have to change some class names/imports as there are new annotations for Running tests in JUnit 5. It won't be that much to integrate (to my understanding), I just have to change a few things here and there in every Java Test classes with Junits, although I am not sure how many there are

**Comment by arnavarora2004** [comment]
he migration to Junit5 requires editing in every single JUnit test class. I finished an entire sub directory myself but dependent on how many classes there are to change, the scope is a bit larger than I expected. There are a some annotation changes needed in every JUnit test class as well as depreciated methods that require some work around, I made a good stride in it but someone with more time on their hands should definitely work on this.
